### PR TITLE
chore(*) upgrade protocol buffers version

### PR DESCRIPTION
**Pull request by jakubdyszkiewicz**
### The problem

In order to upgrade to the newest go-control-plane which will enable upgrade to V3 API, we need to upgrade protobuf to 1.4.x

### Proto API V2
In March 2020, Protobuf team introduced Proto API V2 https://blog.golang.org/protobuf-apiv2 (do not confuse this with proto language version. We are still on Proto 3).
The old API was designed before Go 1.0. There were too many changes they wanted to do so they released the new API.
The new API is available under `google.golang.org/protobuf` go package/module, not `github.com/golang/protobuf`.
HOWEVER, when you change to `github.com/golang/protobuf` 1.4.x, the protobufs are now generated using new code, but they are backward compatibile to API V1. So you can generate protos that are compatible with API V1, V2.

The new proto Go code however has some breaking changes. The new code embeds the new structure called [MessageState](https://godoc.org/google.golang.org/protobuf/internal/impl#MessageState). This new structure contains two "pragmas":
* [DoNotCopy](https://godoc.org/google.golang.org/protobuf/internal/impl#MessageState) that prevents Go proto structure to be "shallow" copied. They did this by embedding empty mutex slice, so go vet complains about this. This means that you cannot pass protos by value, you have to use pointers.
* [DoNotCompare](https://godoc.org/google.golang.org/protobuf/internal/pragma#DoNotCompare). Essentially, if you want to do equal now, you have to use `proto.equal`, simple equal is not enough because the proto structure has a state now.

### Do Not Copy

Why it was introduced? It's because of the state, but why there is a state? It's hard to say. The only explanation I found is that _you are doing something wrong if you are copying protobuf structures_ https://stackoverflow.com/questions/64183794/why-do-the-go-generated-protobuf-files-contain-mutex-locks

This change is the reason why this PR is so big. Our core model looked like this
```
type DataplaneResource struct {
	Meta model.ResourceMeta
	Spec mesh_proto.Dataplane
}
```
then we could initiate it with
```
dpRes := DataplaneResource{}
```
and we had initialized proto structure with default empty proto. No nils, no problems.

The problem is that even that with such design, it is almost impossible to avoid copying. As an example
```
func (t *DataplaneResource) SetSpec(spec model.ResourceSpec) error {
	dataplane, ok := spec.(*mesh_proto.Dataplane)
	if !ok {
		return errors.New("invalid type of spec")
	} else {
		t.Spec = *dataplane // <- here, we just made a copy
		return nil
	}
}
```

There were MANY places in the code that we were passing proto by value and every single one was a problem.

The only solution that I could see is to switch the model to store the pointer
```
type DataplaneResource struct {
	Meta model.ResourceMeta
	Spec *mesh_proto.Dataplane
}
```
This small changes caused massive number of changes. The project assumed that the spec cannot be nil in many places so the simple `DataplaneResource{}` caused many problems.
I added the `NewXResource()` method next to every resource, so it is now a recommended way to start working with the object. However, you can still use struct directly `DataplaneResource{` if you set the `Spec` part.

I was also thinking about adding nil checks in some strategic places that were the most problematic before I fixed all the tests like json marshall / unmarshall, but I feel like it's hiding a potential null that we should not allow to have.

### Do Not Compare

Equal() matcher does not longer work on proto. I introduced new matcher: MatchProto() (following pattern of MatchJSON() and MatchYAML()).
Also we were using DeepEquals in some places in the code which is not allowed now.

### Ignore SA1019

staticcheck complains that we are using deprecated `github.com/golang/protobuf` (V1) instead of `google.golang.org/protobuf` (V2).
I tried to migrate to API V2, but go-control-plane is still on V1 so I saw some interop issues, so I backed off for now by ignoring depracation.

### Proto changes

Proto now serializes differently in some cases. For example when the list is empty, it will omit it so instead of
```
tracing:
  backends: []
```
we will see
```
tracing: {}
```

protoc now complains if we do not add full path into `go_package` so I added this to some protos (System, Zone, Config, kumactl config etc.). I also removed validation files since we are not using them. We were using them with kumactl config but we validate the config ourselves anyways, so it's not needed.

### CODE REVIEW

File: .golangci.yml [hunk 1]
```diff
@@ -17,4 +17,19 @@ run:
...
+      text: "SA1019: package github.com/golang/protobuf/proto is deprecated" # ignore deprecation of proto library. We don't want to migrate yet because go-control-plane is not ready
+    - linters:
+        - staticcheck
+      text: "SA1019: package github.com/golang/protobuf/jsonpb" # ignore deprecation of proto library. We don't want to migrate yet because go-control-plane is not ready
+    - linters:
+        - staticcheck
+      text: "SA1019: proto.MessageName is deprecated" # ignore deprecation of proto library. We don't want to migrate yet because go-control-plane is not ready
+    - linters:
+        - staticcheck
+      text: "SA1019: proto.MessageType is deprecated" # ignore deprecation of proto library. We don't want to migrate yet because go-control-plane is not ready
```
Comment (nickolaev) [inline .golangci.yml:35]: Shall we add `TODO` here so we can find such pieces in the code?

File: app/kumactl/cmd/config/testdata/config-control-planes-remove.13.golden.yaml [hunk 2]
```diff
@@ -1,2 +1 @@
-contexts: []
-controlPlanes: []
+{}
```
Comment (nickolaev) [inline app/kumactl/cmd/config/testdata/config-control-planes-remove.13.golden.yaml:1]: OK, that is interesting now.  Why is this?

File: api/Makefile [hunk 3]
```diff
@@ -21,7 +21,7 @@ endef
 
 PROTOC_VERSION := 3.6.1
 PROTOC_PGV_VERSION := v0.3.0-java.0.20200311152155-ab56c3dd1cf9 # Envoy API from master needs changes from master in protoc-gen-validate. Switch to stable once it's released
-GOLANG_PROTOBUF_VERSION := v1.3.2
+GOLANG_PROTOBUF_VERSION := v1.4.3
```
Comment (nickolaev) [inline api/Makefile:24]: Shall we use this PR to also bump protoc version?

File: pkg/core/resources/apis/mesh/dataplane_overview.go [hunk 4]
```diff
@@ -98,14 +102,14 @@ func NewDataplaneOverviews(dataplanes DataplaneResourceList, insights DataplaneI
 	for _, dataplane := range dataplanes.Items {
 		overview := DataplaneOverviewResource{
 			Meta: dataplane.Meta,
-			Spec: mesh_proto.DataplaneOverview{
-				Dataplane:        proto.Clone(&dataplane.Spec).(*mesh_proto.Dataplane),
+			Spec: &mesh_proto.DataplaneOverview{
+				Dataplane:        dataplane.Spec,
```
Comment (nickolaev) [inline pkg/core/resources/apis/mesh/dataplane_overview.go:106]: OK, so this is just an overview, so I guess Spec is immutable here, but how do we guarantee safety now that we use pointers everywhere?

**Review by nickolaev — APPROVED** [review]
Overall very exciting. Approving here, please check my remarks though.

File: api/Makefile (hunk 3 above)
Comment (nickolaev) [reply api/Makefile:24]: Also https://github.com/envoyproxy/protoc-gen-validate/releases

File: app/kumactl/cmd/config/testdata/config-control-planes-remove.13.golden.yaml (hunk 2 above)
Comment (jakubdyszkiewicz) [reply app/kumactl/cmd/config/testdata/config-control-planes-remove.13.golden.yaml:1]: "Proto changes" in the readme

Comment (nickolaev) [reply app/kumactl/cmd/config/testdata/config-control-planes-remove.13.golden.yaml:1]: Nevermind, I just read the PR description to the end.

File: pkg/core/resources/model/rest/resource_test.go [hunk 5]
```diff
@@ -109,28 +110,23 @@ var _ = Describe("Rest Resource", func() {
...
-							Mesh: "demo",
-							Name: "two",
-						},
-						Spec: &sample_proto.TrafficRoute{
-							Path: "/another",
-						},
-					}),
-				)
+				Expect(rs.Items).To(HaveLen(2))
+				Expect(rs.Items[0].Meta).To(Equal(rest.ResourceMeta{
```
Comment (lobkovilya) [inline pkg/core/resources/model/rest/resource_test.go:114]: Is the order of items guaranteed?

File: api/Makefile (hunk 3 above)
Comment (nickolaev) [reply api/Makefile:24]: I did change it on to of yours in this branch just to check CI
https://github.com/kumahq/kuma/tree/chore/upgrade-proto-protoc

File: .golangci.yml (hunk 1 above)
Comment (jakubdyszkiewicz) [reply .golangci.yml]: added

File: pkg/core/resources/model/rest/resource_test.go (hunk 5 above)
Comment (lobkovilya) [reply pkg/core/resources/model/rest/resource_test.go:114]: because `ConsistOf` didn't check the order

Comment (jakubdyszkiewicz) [reply pkg/core/resources/model/rest/resource_test.go:114]: yes, the order is guaranteed

**Review by lobkovilya — APPROVED** [review]

File: api/Makefile (hunk 3 above)
Comment (jakubdyszkiewicz) [reply api/Makefile:24]: good point, upgraded

File: pkg/core/resources/apis/mesh/dataplane_overview.go (hunk 4 above)
Comment (jakubdyszkiewicz) [reply pkg/core/resources/apis/mesh/dataplane_overview.go:106]: Yeah... we cannot really :( that's the huge downside of having pointers.

Here I noticed that the cloning is unnecessary so I removed it.

**Pull request merged**